fedimint-client/src/transaction/builder.rs

```diff
@@ -420,14 +450,60 @@ impl TransactionBuilder {
             signatures: TransactionSignature::NaiveMultisig(signatures),
         };
 
-        let states = input_sms
+        let input_states = self
+            .inputs
             .into_iter()
             .enumerate()
-            .chain(output_sms.into_iter().enumerate())
-            .flat_map(|(idx, state_gen)| state_gen(txid, idx as u64))
-            .collect::<Vec<_>>();
+            .flat_map(|(bundle_idx, bundle)| {
+                let input_idxs = input_idx_to_bundle_idx
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(input_idx, input_bundle_idx)| {
+                        (*input_bundle_idx == bundle_idx).then_some(input_idx as u64)
+                    })
+                    .fold(None, |cur: Option<(u64, u64)>, idx| {
+                        Some(cur.map_or((idx, idx), |cur| (cur.0.min(idx), cur.1.max(idx))))
+                    });
```

Comment:
There has to be a better way to recover input index ranges :laughing: This code was fairly hard to understand unfortunately. Maybe accepting mutation when adding inputs is the price we have to pay for not having to reconstruct the ranges like this.

Also, the algorithm scales quadratically with the number of inputs/bundles (likely not a real problem, but meh).

Reply:
I don't find this code offensive. Sure, it's well... immediately obvious, but I can add some comments, and the general model there is actually quite simple.

Reply:
I actually found an elegant refactoring that makes it easier to look at.